hal/sai_api.h

```diff
@@ -0,0 +1,195 @@
+ * - `sai_xfer()` returns false if there's no sample in the FiFo ;
+ * - `sai_xfer()` if `psample` is NULL : it pops 1 sample from the FiFo and returns true ;
+ * - `sai_xfer()` if `psample` is not NULL : it pops 1 sample from the FiFo, stores it to the address
+ *   pointed by `psample`,  and returns true.
+ *
+ * if the device is a *transmitter* :
+ * - `sai_xfer()` returns false if the `sai_t` object is NULL ;
+ * - `sai_xfer()` returns false if the fifo is full and `*psample` could not be pushed ;
+ * - `sai_xfer()` if `psample` is NULL : it pushes one '0' sample to the FiFo and returns true ;
+ * - `sai_xfer()` if `psample` is not NULL : it pushes the pointed sample to the FiFo and returns true.
+ *
+ * # Undefined behaviours
+ * - Calling any function other than `sai_init()` before the initialization of the SAI ;
+ * - Calling any function other than `sai_init()` after calling `sai_free()`.
+ *
+ * # other requirements
+ * A target must also define these elements to allow tests to be run.
+ * - `#define SAI_DEFAULT_SAMPLE_RATE (xxxxxU)` ;
+ * - Pins for 2 SAI/I²S interface including MCLK, BCLK, WCLK and SD named respectively
+ *   - SAI_A_MCLK, SAI_A_BCLK, SAI_A_WCLK and SAI_A_SD ;
+ *   - SAI_B_MCLK, SAI_B_BCLK, SAI_B_WCLK and SAI_B_SD.
+ *
+ * @note
+ * A transceiver supporting async rx/tx should be considered as 2 different peripherals :
+ * - one read-only
+ * - one write-only
+ * The first allocated channel may or may not limit the second one's feature.
+ * eg:
+ * In a peripheral that supports async rx/tx but requires format to be the same,
+ * the first allocated instance will set the format and tie the second one to this format.
+ *
+ * @{
+ */
+ 
+/**
+ * \defgroup hal_sai_tests sai hal tests
+ * The sai HAL tests ensure driver conformance to defined behaviour.
+ *
+ * To run the sleep hal tests use the command:
+ *
+ *     mbed test -t <toolchain> -m <target> -n tests-mbed_hal-sai*
+ *
+ */
+
+
+/** Defines frame format. */
+typedef struct sai_format_s {
+  bool          bclk_polarity;  /**< true for Active high */
+  bool          wclk_polarity;  /**< true for Active high */
+  bool          ws_delay;       /**< true to toggle ws one bit earlier than the frame */
+  uint8_t       ws_length;      /**< ws assertion length from 1bclk to word length. */
+
+  uint8_t       frame_length;   /**< Frame length in word count [1; 32] */
+  uint32_t      word_mask;      /**< Mask on frame for used word (bitfield) */
+  uint8_t       word_length;    /**< Word length in bits [1; 32] */
+  uint8_t       data_length;    /**< Data length within the word. This must <= to word_length. */
+  bool          lsb_first;      /**< true to send lsb first */
+  bool          aligned_left;   /**< true to align Left */
+  uint8_t       bit_shift;      /**< sample bit shift distance from its alignment point. */
+} sai_format_t;
+
+/** Defines input mclk source */
+typedef enum sai_clock_source_e {
+  SAI_CLOCK_SOURCE_Internal,
```

Comment:
Does'nt that cause confusion between enums & macros ?

Reply:
Definitely not the mix of both. Looking at HAL enums, we do uppercase in most cases (serial has this different, for instance `FlowControl` enum `FlowControlNone` - serial_api is the only exception). This might need to be added to our style 😕

Reply:
if it can help for the record, I used to do like this for enums.
`UPPERCASE_ENUM_NAME_Lowercase_but_the_first_for_enum_variant`
This makes a clear distinction between FULLY_CAPITALIZED_MACROS and enum variants.

Do you want that "fixed" to either full caps or CamelCase ?
Can this also be postponed to the next commit after this is merged ?

Reply:
being consistent, then `FULLY_CAPITALIZED_MACROS`, fine to come in as separate patch but prefer to have it here :-) should be simple to refactor tests quickly